sql: add buffer and scanBuffer nodes

yuzefovich opened the pull request:
Adds bufferNode that consumes its input, stores all the rows in a
buffer, and then proceeds on passing the rows through. The buffer
can be iterated over multiple times using scanBuffer node that is
referencing a single bufferNode.

Fixes: #37050.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/37137)
<!-- Reviewable:end -->

yuzefovich commented:
I left several comments on things I'm not sure about.

Also, I can't seem to find an example of how we unit-test local plan nodes, any pointers?

RaduBerinde reviewed (commented):
Sorry for the late review. We don't typically do unit tests for plan nodes, it would be very tedious to set up. We usually test them through logic tests. It's fine to hold off on any tests until we make use of the nodes.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37137#-:-LdiU9or-Uh8_6MiqZfr:bwxxev5)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @justinj, @RaduBerinde, and @yuzefovich)

___
*[pkg/sql/buffer.go, line 28 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37137#-LdiTje9-a7cr0G_uDAT:-LdiTje9-a7cr0G_uDAU:b-u2suq1) ([raw file](https://github.com/cockroachdb/cockroach/blob/17b958d6836648624d84c5a63311d81a883a5a3f/pkg/sql/buffer.go#L28)):*
> ```Go
> // After the input has been fully consumed, it proceeds on passing the rows
> // through. The buffer can be iterated over multiple times.
> // TODO(yuzefovich): is this buffering all rows at once desirable?
> ```

Yes, it's kind of required in general. Part of the point is that we need to consume all input rows (because the input can be a mutation) even if the upper node doesn't need it.
___
*[pkg/sql/buffer.go, line 29 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37137#-LdiU0EY9Z52WsDU3_Gl:-LdiU0EZ7kQ3xy947IKK:b-tfw1hi) ([raw file](https://github.com/cockroachdb/cockroach/blob/17b958d6836648624d84c5a63311d81a883a5a3f/pkg/sql/buffer.go#L29)):*
> ```Go
> // through. The buffer can be iterated over multiple times.
> // TODO(yuzefovich): is this buffering all rows at once desirable?
> // TODO(yuzefovich): current version supports only a single scanBufferNode at a
> ```

We will need to be able to have multiple `scanBufferNode` that acts like independent scans.
___
*[pkg/sql/buffer.go, line 52 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37137#-LdiTr1JEAwdf8o1naOU:-LdiTr1JEAwdf8o1naOV:b-bxrkuz) ([raw file](https://github.com/cockroachdb/cockroach/blob/17b958d6836648624d84c5a63311d81a883a5a3f/pkg/sql/buffer.go#L52)):*
> ```Go
> 
> func (n *bufferNode) Next(params runParams) (bool, error) {
> 	if !n.inputDone {
> ```

It feels like this should happen in `startExec`. If the parent node ends up not needing any rows, we still need to execute the encapsulated plan fully.
___
*[pkg/sql/buffer.go, line 96 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37137#-LdiVmx9CPR3yzisoGng:-LdiVmx9CPR3yzisoGnh:b-xj1lvu) ([raw file](https://github.com/cockroachdb/cockroach/blob/17b958d6836648624d84c5a63311d81a883a5a3f/pkg/sql/buffer.go#L96)):*
> ```Go
> // referencing. The bufferNode can be iterated over multiple times, however, a
> // new scanBufferNode is needed.
> type scanBufferNode struct {
> ```

Add a comment that scanBufferNode can only start execution after the bufferNode finishes its execution completely.


<!-- Sent from Reviewable.io -->

yuzefovich committed:
sql: add buffer and scanBuffer nodes

Adds bufferNode that consumes its input, stores all the rows in a
buffer, and then proceeds on passing the rows through. The buffer
can be iterated over multiple times using scanBuffer node that is
referencing a single bufferNode.

Release note: None

yuzefovich reviewed (commented):
Ok, makes sense, thanks.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37137#-:-LdoTa2a0EU2pMIeezKt:bsqt64z)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @justinj, and @RaduBerinde)

___
*[pkg/sql/buffer.go, line 28 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37137#-LdiTje9-a7cr0G_uDAT:-LdoTu8bEI081OL3ENhq:bmye5ai) ([raw file](https://github.com/cockroachdb/cockroach/blob/17b958d6836648624d84c5a63311d81a883a5a3f/pkg/sql/buffer.go#L28)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Yes, it's kind of required in general. Part of the point is that we need to consume all input rows (because the input can be a mutation) even if the upper node doesn't need it.
</blockquote></details>

Ok, I see, thanks.
___
*[pkg/sql/buffer.go, line 29 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37137#-LdiU0EY9Z52WsDU3_Gl:-LdobJ_JEfbEK28DtId2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/17b958d6836648624d84c5a63311d81a883a5a3f/pkg/sql/buffer.go#L29)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

We will need to be able to have multiple `scanBufferNode` that acts like independent scans.
</blockquote></details>

Done.
___
*[pkg/sql/buffer.go, line 52 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37137#-LdiTr1JEAwdf8o1naOU:-LdoUjwU-YQcuB_WIEpY:b-opbltv) ([raw file](https://github.com/cockroachdb/cockroach/blob/17b958d6836648624d84c5a63311d81a883a5a3f/pkg/sql/buffer.go#L52)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

It feels like this should happen in `startExec`. If the parent node ends up not needing any rows, we still need to execute the encapsulated plan fully.
</blockquote></details>

But is it safe to be `Next`ing the input in `startExec`? My understanding is as follows: say we have the following tree `valuesNode -> bufferNode -> insertNode`, `startExec` will be called on `insertNode` first, then on `bufferNode` before the call on `valuesNode` due to depth-first traversal in `plan.go/startExec`. So if we try buffering all the rows in `startExec`, we'll be calling `Next` on `valuesNode` before it has been started. Possibly, I'm missing something, or there needs to be other changes to accommodate buffering within `startExec`.
___
*[pkg/sql/buffer.go, line 96 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37137#-LdiVmx9CPR3yzisoGng:-LdoWAJZ8AO5WeEpdAw4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/17b958d6836648624d84c5a63311d81a883a5a3f/pkg/sql/buffer.go#L96)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Add a comment that scanBufferNode can only start execution after the bufferNode finishes its execution completely.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37137)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @justinj, and @yuzefovich)

___
*[pkg/sql/buffer.go, line 52 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37137#-LdiTr1JEAwdf8o1naOU:-LdpXJxv6CL1tR856FBh:b-10brid) ([raw file](https://github.com/cockroachdb/cockroach/blob/17b958d6836648624d84c5a63311d81a883a5a3f/pkg/sql/buffer.go#L52)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

But is it safe to be `Next`ing the input in `startExec`? My understanding is as follows: say we have the following tree `valuesNode -> bufferNode -> insertNode`, `startExec` will be called on `insertNode` first, then on `bufferNode` before the call on `valuesNode` due to depth-first traversal in `plan.go/startExec`. So if we try buffering all the rows in `startExec`, we'll be calling `Next` on `valuesNode` before it has been started. Possibly, I'm missing something, or there needs to be other changes to accommodate buffering within `startExec`.
</blockquote></details>

I think it's post-order traversal https://github.com/cockroachdb/cockroach/blob/147489fe1793f9b1170f95ab353caf1b199cb83b/pkg/sql/plan.go#L480


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37137)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @justinj)

___
*[pkg/sql/buffer.go, line 52 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37137#-LdiTr1JEAwdf8o1naOU:-LdpYjeA4Kp1T6_3jpOl:bjx4uon) ([raw file](https://github.com/cockroachdb/cockroach/blob/17b958d6836648624d84c5a63311d81a883a5a3f/pkg/sql/buffer.go#L52)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I think it's post-order traversal https://github.com/cockroachdb/cockroach/blob/147489fe1793f9b1170f95ab353caf1b199cb83b/pkg/sql/plan.go#L480
</blockquote></details>

Done. You're right - I was confused (I was thinking that in the example `valuesNode` was the root of the tree while it is a leaf and `insertNode` is the root, right?).


<!-- Sent from Reviewable.io -->

jordanlewis reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37137)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @justinj, and @yuzefovich)

___
*[pkg/sql/buffer.go, line 33 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37137#-LeFQcsx8XusneNFOLXt:-LeFQcsx8XusneNFOLXu:by1u35) ([raw file](https://github.com/cockroachdb/cockroach/blob/c89548c980695910ec45be5c879a71e88503b40f/pkg/sql/buffer.go#L33)):*
> ```Go
> 	// TODO(yuzefovich): should the buffer be backed by the disk? If so, the
> 	// comments about TempStorage suggest that it should be used by DistSQL
> 	// processors, but this node is local.
> ```

I think eventually.
___
*[pkg/sql/buffer.go, line 73 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37137#-LeFQbu391A089EhEj1W:-LeFQbu391A089EhEj1X:b-nbxz02) ([raw file](https://github.com/cockroachdb/cockroach/blob/c89548c980695910ec45be5c879a71e88503b40f/pkg/sql/buffer.go#L73)):*
> ```Go
> }
> 
> // TODO(yuzefovich): does this need to have some special behavior?
> ```

This looks fine to me.
___
*[pkg/sql/buffer.go, line 103 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37137#-LeFQTtS5n20xD2FA18n:-LeFQTtT28-haaNKUibC:b16vu4o) ([raw file](https://github.com/cockroachdb/cockroach/blob/c89548c980695910ec45be5c879a71e88503b40f/pkg/sql/buffer.go#L103)):*
> ```Go
> }
> 
> // Note that scanBufferNode does not close the corresponding to it bufferNode.
> ```

nit: this sentence doesn't make too much sense.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
@RaduBerinde please let me know if this is good to be merged now (with likely modifications later).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37137#-:-LeNLHerBWyjguLv3y93:b2sj72u)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @justinj)

___
*[pkg/sql/buffer.go, line 33 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37137#-LeFQcsx8XusneNFOLXt:-LeNKYxpAjrafog5glle:b-2a2lcu) ([raw file](https://github.com/cockroachdb/cockroach/blob/c89548c980695910ec45be5c879a71e88503b40f/pkg/sql/buffer.go#L33)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I think eventually.
</blockquote></details>

Ok, I'll leave this TODO for now then.
___
*[pkg/sql/buffer.go, line 73 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37137#-LeFQbu391A089EhEj1W:-LeNKlPH2twG49TEp_wJ:b-54avio) ([raw file](https://github.com/cockroachdb/cockroach/blob/c89548c980695910ec45be5c879a71e88503b40f/pkg/sql/buffer.go#L73)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

This looks fine to me.
</blockquote></details>

Ok, thanks.
___
*[pkg/sql/buffer.go, line 103 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37137#-LeFQTtS5n20xD2FA18n:-LeNKoFf12f0As6ehOd3:bss4a2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/c89548c980695910ec45be5c879a71e88503b40f/pkg/sql/buffer.go#L103)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

nit: this sentence doesn't make too much sense.
</blockquote></details>

Removed.


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Now that I think of it more, I think we will need both kinds of semantics:
 - a buffer node which passes through all rows immediately. It can assume that the parent will read through all the rows. This will be used as input to mutations.
 - a buffer node which first buffers all rows upfront. This will be used for CTEs. In this case the parent node will not need the rows.TestImportCSVStmt

I am not sure if we want them to be separate nodes or not. It might be cleaner if they're separate, especially since in the second case we don't need it to return the rows.

RaduBerinde commented:
The second case above would be used by a `withNode` which buffers the left side and then runs the right side. From that standpoint, maybe we don't need a separate kind of buffer node - the `withNode` could just read all rows from a bufferNode on the left side (and throw them away) before executing the right side.

yuzefovich reviewed (commented):
I made adjustments to support the first case, and your point about discarding the rows in the second case makes sense (I think, actually, `withNode` could be just `Next`ing the `bufferNode` without explicitly reading the rows with `Values`).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37137#-:-Leso8uk5mZ29abjPeep:b-3d4yug)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @justinj)



<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Thanks!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37137#-:-Lesr1Xx6pgyJeBUfAT2:b-g2a5fb)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis and @justinj)



<!-- Sent from Reviewable.io -->

yuzefovich commented:
TFTRs!

bors r+
